_posts/2017-11-29-Express-Reloadable-Update.markdown

```diff
@@ -0,0 +1,47 @@
+
+<!-- more -->
+
+It's common to share NPM packages across projects, and oftentimes packages are developed in parallel. Package `A` depends on `B`, but `B` has a bug and you don't want to have to republish (and reinstall) the package in order to see changes made locally. `yarn link` (or `npm link`) was developed for instances like this and while it works great for stop and start processes where boot time is quick, it falls short if the development environment takes a while to load. In UI-rich environments like [Positron](https://github.com/artsy/positron) (our Editorial "Writer" app) and [Force](https://github.com/artsy/force), each boot would come at a high time-cost due to upfront compilation of assets. Tools like [nodemon](https://github.com/remy/nodemon) would automatically stop and start our server process when assets changed but that still didn't alleviate slow iteration times.
+
+To recap from a [previous post](artsy.github.io/blog/2017/09/05/Modernizing-Force/), [@artsy/express-reloadable](https://github.com/artsy/express-reloadable) allows devs to immediately see changes to running Express.js app code:
+
+```javascript
+import express from 'express'
+import { createReloadable, isDevelopment } from '@artsy/express-reloadable'
+
+const app = express()
+
+if (isDevelopment) {
+  const mountAndReload = createReloadable(app, require)
+  mountAndReload('./api')
+}
+```
+
+Changes made within the `api` folder are detected and hot-swapped in, and all that's required is a new http request. However, we found an exception while building out Artsy's new [editorial pages](https://www.artsy.net/article/artsy-editorial-midwest-made-artists-mike-kelley-jim-shaw), which involved [sharing Reaction components](https://github.com/artsy/positron/blob/master/client/apps/edit/components/content/section_tool/index.jsx#L11) between Positron [and Force](https://github.com/artsy/force/blob/master/desktop/apps/article/components/InfiniteScrollArticle.js#L9). Even though we ran `yarn link @artsy/reaction` in each consumer app, changes would not be detected and so we had to do a full restart.
+
+To address this, a new `watchModules` feature [was added](https://github.com/artsy/express-reloadable/pull/3):
+
+```javascript
+mountAndReload('./api', {
+  watchModules: [
+    '@artsy/reaction',
+  ]
+})
+```
+
+Similar to how files in `api/` are hot-swapped in on change, symlinked NPM modules placed in the `watchModules` array will now be reloaded, too.
```

Comment:
This could probably do with a _tiny_ summary: like 
"With express-reloadable we dropped our dev cycle from 6s to 0.2s, except when working in libraries - now with `watchModules` that's down to an average 0.3s" or something

Reply:
More like 30 seconds :)

Reply:
lolz

Reply:
Just timed it out of curiosity -- 43 sec... updating 👍

Reply:
Holy moly! Is that us just iterating over all the file entries in the cache?

Reply:
@alloy no, not exactly 😄  to clarify, when we boot force we mount a bunch of sub-apps and then kick off `browserify-dev-middleware` which then compiles our assets. Prior to adding in `express-reloadable`, whenever a server-side change was made we'd have to stop the node.js process and restart, wait, and reload the browser or make a new request before we could see changes. Now that `express-reloadable` is in place we watch our apps folder and whenever a change takes place it will flush the cache (which is pretty instant) allowing for fresh code to run.

Reply:
Ohhhhhhhhhh… is this about the time it took to reload _before_ we flushed the cache?

Reply:
@alloy, yup. This is like raw, required boot time independent of all else

Reply:
Ahh 👌